src/main/java/de/tum/in/www1/artemis/config/MetricsBean.java

```diff
@@ -113,13 +127,108 @@ private void registerWebsocketMetrics() {
+
+                Gauge.builder("artemis.scheduled.exercises.release.count", () -> this.getUpcomingReleasedExercisesCount(range, exerciseType)).strongReference(true)
+                        .tag("range", String.valueOf(range)).tag("exerciseType", exerciseType.toString()).description("Number of exercises starting within the next minutes")
+                        .register(meterRegistry);
+
+                Gauge.builder("artemis.scheduled.exercises.release.student_multiplier", () -> this.getUpcomingReleasedExercisesCountWithStudentMultiplier(range, exerciseType))
+                        .strongReference(true).tag("range", String.valueOf(range)).tag("exerciseType", exerciseType.toString())
+                        .description("Number of exercises starting within the next minutes multiplied with students in the course").register(meterRegistry);
+            }
+
+            Gauge.builder("artemis.scheduled.exams.due.count", range, this::getUpcomingEndingExamCount).strongReference(true).tag("range", String.valueOf(range))
+                    .description("Number of exams ending within the next minutes").register(meterRegistry);
+
+            Gauge.builder("artemis.scheduled.exams.due.student_multiplier", range, this::getUpcomingEndingExamCountWithStudentMultiplier).strongReference(true)
+                    .tag("range", String.valueOf(range)).description("Number of exams ending within the next minutes multiplied with students in the course")
+                    .register(meterRegistry);
+
+            Gauge.builder("artemis.scheduled.exams.release.count", range, this::getUpcomingStartingExamCount).strongReference(true).tag("range", String.valueOf(range))
+                    .description("Number of exams starting within the next minutes").register(meterRegistry);
+
+            Gauge.builder("artemis.scheduled.exams.release.student_multiplier", range, this::getUpcomingStartingExamCountWithStudentMultiplier).strongReference(true)
+                    .tag("range", String.valueOf(range)).description("Number of exams starting within the next minutes multiplied with students in the course")
+                    .register(meterRegistry);
+        }
+    }
+
     private static double extractWebsocketUserCount(WebSocketHandler webSocketHandler) {
         if (webSocketHandler instanceof SubProtocolWebSocketHandler subProtocolWebSocketHandler) {
             return subProtocolWebSocketHandler.getStats().getWebSocketSessions();
         }
         return -1;
     }
 
+    private double getUpcomingDueExercisesCount(int minutes, ExerciseType exerciseType) {
+        SecurityUtils.setAuthorizationObject();
+        var now = ZonedDateTime.now();
+        var endDate = ZonedDateTime.now().plusMinutes(minutes);
+        return exerciseRepository.findAllExercisesWithCurrentOrUpcomingDueDateWithinTimeRange(now, endDate).stream().filter(e -> e.getExerciseType() == exerciseType).count();
```

Comment:
Can we directly count as part of the SQL query instead? That would avoid the construction of a full list of exercise objects to just then immediately throw them away again.

Similarly for the other queries. For the queries that count users a join that counts and sums up could maybe work as well.

Reply:
I tried to do that but had issues filtering by the type of the exercise since the `getExerciseType()`-method is only available in the Java Objects (but not persisted in the database). In the database, the discriminator is stored instead and I'm not sure how to filter by that.
Do you know how this could be done?

Reply:
Hmm, getting the discriminator from the exercise type works:
```java
enum ExerciseType {
    // […]

    public String getDiscriminator() {
        final DiscriminatorValue value = getExerciseClass().getAnnotation(DiscriminatorValue.class);
        return value.value();
    }

    private Class<? extends Exercise> getExerciseClass() {
        return switch (this) {
            case TEXT -> TextExercise.class;
            case PROGRAMMING -> ProgrammingExercise.class;
            case MODELING -> ModelingExercise.class;
            case FILE_UPLOAD -> FileUploadExercise.class;
            case QUIZ -> QuizExercise.class;
        };
    }
}
```

This could then maybe be used in the queries for the `TYPE` filter, e.g.
```java
    @Query("""
            SELECT e FROM Exercise e
            WHERE e.course.testCourse = FALSE
            	AND e.dueDate >= :#{#now}
            	AND e.dueDate <= :#{#maxDate}
            	AND TYPE(e) = :#{#exerciseType}
            ORDER BY e.dueDate ASC
            """)
    Set<Exercise> findAllExercisesWithCurrentOrUpcomingDueDateWithinTimeRange(@Param("now") ZonedDateTime now, @Param("maxDate") ZonedDateTime maxDate, @Param("exerciseType") String exerciseTypeDiscriminator);

```

Reply:
@b-fein Thanks for your suggestion, I had issues with `TYPE(e)` but found another solution using `e.class` (see https://stackoverflow.com/a/5584207/3802758 or https://docs.jboss.org/hibernate/core/3.6/reference/en-US/html/queryhql.html#queryhql-identifier-property).

Could you check if this solution is fine for you as well?

![image](https://user-images.githubusercontent.com/5084100/205161449-f7bb8853-f042-47c4-9a19-4a527841e441.png)

Reply:
Nice, looks good! Of course using another operator (that probably gets converted to the same SQL anyway) is fine.